parsers/marc.go

```diff
@@ -0,0 +1,158 @@
+	contributor []string
+	url         []string
+	subject     []string
+	isbn        []string
+}
+
+const marcRules = "../fixtures/marc_rules.json"
+
+// Rules defines where the rules are in JSON
+type Rules struct {
+	Field     string `json:"field"`
+	Tag       string `json:"tag"`
+	Subfields string `json:"subfields"`
+}
+
+// RetrieveRules for parsing MARC
+func RetrieveRules(rulefile string) ([]*Rules, error) {
+	// Open the file.
+	file, err := os.Open(rulefile)
+	if err != nil {
+		return nil, err
+	}
+
+	// Schedule the file to be closed once
+	// the function returns.
+	defer file.Close()
+
+	// Decode the file into a slice of pointers
+	// to Feed values.
+	var rules []*Rules
+	err = json.NewDecoder(file).Decode(&rules)
+
+	// We don't need to check for errors, the caller can do this.
+	return rules, err
+}
+
+// Process kicks off the MARC processing
+func Process(rulesfile string) {
+
+	var records []record
+
+	rules, err := RetrieveRules(rulesfile)
+	if err != nil {
+		spew.Dump(err)
+		return
+	}
+
+	// loop over all records
+	count := 0
+	for {
+		record, err := marc21.ReadRecord(os.Stdin)
+
+		// if we get an error, log it
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			log.Println("An error occured processing the", count, "record.")
+			log.Fatal(err)
+		}
+
+		count++
+
+		// we probably don't want to make this in memory representation of the
+		// combined data but instead will probably want to open a JSON file for
+		// writing at the start of the loop, write to it on each iteration, and
+		// close it when we are done. Or something. Channels?
+		// For now I'm just throwing everything into a slice and dumping it because
+		// :shrug:
+		records = append(records, marcToRecord(record, rules))
+	}
+	spew.Dump(records)
+	log.Println("Processed ", count, "records")
+}
+
+func marcToRecord(marcRecord *marc21.Record, rules []*Rules) record {
+	r := record{}
+
+	r.identifier = marcRecord.Identifier()
+
+	// main entry
+	rule := rules[0]
+	r.title = concatSubfields(rule.Tag, []byte(rule.Subfields), marcRecord)[0]
+
+	// author
+	r.author = toRecord(r.author, rules[1], marcRecord)
+
+	// contributors
+	r.contributor = toRecord(r.contributor, rules[2], marcRecord)
+
+	// urls
+	r.url = toRecord(r.url, rules[3], marcRecord)
+
+	// subjects
+	r.subject = toRecord(r.subject, rules[4], marcRecord)
+	r.subject = toRecord(r.subject, rules[5], marcRecord)
+	r.subject = toRecord(r.subject, rules[6], marcRecord)
+	r.subject = toRecord(r.subject, rules[7], marcRecord)
+
+	//isbn
+	r.isbn = toRecord(r.isbn, rules[8], marcRecord)
+	return r
+}
+
+func toRecord(field []string, rule *Rules, marcRecord *marc21.Record) []string {
+	field = append(field, concatSubfields(rule.Tag, []byte(rule.Subfields), marcRecord)...)
+	return field
+}
+
+// takes a mark field tag and subfields of interest for a supplied marc record and returns them concatenated
+func concatSubfields(marcfield string, subfields []byte, marcrecord *marc21.Record) []string {
```

Comment:
I'm not sure this is going to work. For example, if we have multiple 650 fields we will want to represent these as an array of fields in Elasticsearch, not all lumped together in a single string.

Reply:
This returns an array of strings though...

Reply:
Let me see if we have sample records that confirm this is working as intended.

Reply:
The `record1` fixture has 5 650 fields in the record.
```
=650  \0$aArithmetic$xJuvenile poetry.
=650  \0$aChildren's poetry, American.
=650  \1$aArithmetic$xPoetry.
=650  \1$aAmerican poetry.
=650  \1$aVisual perception.
```

Each ends up as a separate subject:


```
go run . < fixtures/record1.mrc

([]marc.record) (len=1 cap=1) {
 (marc.record) {
  identifier: (string) (len=12) "   92005291 ",
  title: (string) (len=12) "Arithmetic /",
  author: ([]string) (len=1 cap=1) {
   (string) (len=26) "Sandburg, Carl, 1878-1967."
  },
  contributor: ([]string) (len=1 cap=1) {
   (string) (len=15) "Rand, Ted, ill."
  },
  url: ([]string) <nil>,
  subject: ([]string) (len=5 cap=5) {
   (string) (len=27) "Arithmetic Juvenile poetry.",
   (string) (len=28) "Children's poetry, American.",
   (string) (len=18) "Arithmetic Poetry.",
   (string) (len=16) "American poetry.",
   (string) (len=18) "Visual perception."
  },
  isbn: ([]string) (len=1 cap=1) {
   (string) (len=12) "0152038655 :"
  }
 }
}

```

Reply:
I think we need a better name for this method.

Reply:
Ah, my bad. You are correct.